// @shopify/hydrogen/dist/esnext/utilities/log/log.js
import { yellow, red, green, italic, lightBlue } from 'kolorist';
import { getTime } from '../timing.js';
import { parseUrl } from './utils.js';
const defaultLogger = {
    trace(context, ...args) {
        // Re-enable following line to show trace debugging information
        // console.log(context.id, ...args);
    },
    debug(context, ...args) {
        console.log(...args);
    },
    warn(context, ...args) {
        console.warn(yellow('WARN: '), ...args);
    },
    error(context, error, ...extra) {
        const url = context ? ` ${context.url}` : '';
        if (error instanceof Error) {
            console.error(red(`Error processing route:${url}\n${error.stack}`));
        }
        else {
            console.error(red(`Error:${url} ${error}`));
        }
    },
    fatal(context, ...args) {
        console.error(red('FATAL: '), ...args);
    },
    options: () => ({}),
};
let currentLogger = defaultLogger;
function doLog(method, request, ...args) {
    try {
        const maybePromise = currentLogger[method](request, ...args);
        if (maybePromise instanceof Promise) {
            request?.ctx?.runtime?.waitUntil?.(maybePromise.catch((e) => {
                const message = e instanceof Error ? e.stack : e;
                defaultLogger.error(`Promise error from the custom logging implementation for logger.${method} failed:\n${message}`);
            }));
        }
    }
    catch (e) {
        const message = e instanceof Error ? e.stack : e;
        defaultLogger.error(`The custom logging implementation for logger.${method} failed:\n${message}`);
    }
}
export function getLoggerWithContext(context) {
    return {
        trace: (...args) => doLog('trace', context, ...args),
        debug: (...args) => doLog('debug', context, ...args),
        warn: (...args) => doLog('warn', context, ...args),
        error: (...args) => doLog('error', context, ...args),
        fatal: (...args) => doLog('fatal', context, ...args),
        options: () => currentLogger.options(),
    };
}
export const log = getLoggerWithContext({});
export function setLogger(config) {
    if (!config) {
        currentLogger = defaultLogger;
        return;
    }
    const options = {};
    currentLogger = { ...defaultLogger, ...config, options: () => options };
    for (const key of Object.keys(config)) {
        if (!(key in defaultLogger)) {
            delete currentLogger[key];
            options[key] = config[key];
        }
    }
}
const SERVER_RESPONSE_MAP = {
    str: 'streaming SSR',
    rsc: 'Server Components',
    ssr: 'buffered SSR',
};
export function logServerResponse(type, request, responseStatus, didError) {
    const log = getLoggerWithContext(request);
    const coloredResponseStatus = responseStatus >= 500
        ? red(responseStatus)
        : responseStatus >= 400
            ? yellow(responseStatus)
            : responseStatus >= 300
                ? lightBlue(responseStatus)
                : green(responseStatus);
    const fullType = SERVER_RESPONSE_MAP[type] || type;
    const styledType = italic(fullType.padEnd(17));
    const paddedTiming = ((getTime() - request.time).toFixed(2) + ' ms').padEnd(10);
    const url = parseUrl(type, request.url);
    log.debug(`${request.method} ${styledType} ${coloredResponseStatus} ${didError || responseStatus >= 400 ? red('error') : green('ok   ')} ${paddedTiming} ${url}`);
}
